Different behavior between Quarkus and Wildfly distribution: SQLException on POST returns HTTP 201 instead 500

### Describe the bug

Hi,

if a component with a name longer than 255 is created through REST API, Keycloak 17 returns 201 Created instead 500 Internal Server Error.

This bug exists only if Keycloak is running with Quarkus. If Keycloak runs with Wildfly, the REST API correctly returns 500 instead 201.

The bahaivor between Quarkus and Wildfly distribution is inconstant. 

### Version

17.0.0

### Expected behavior

Keycloak returns 500 Internal Server Error, because the component can't create.

curl against `quay.io/keycloak/keycloak:17.0.0-legacy`

```
curl http://localhost:8080/auth/admin/realms/master/components -H "Authorization: Bearer" -H "Content-Type: application/json" -d '{"id":null,"name":"em3eedoh8biuph8aiwo0oe2Uyaexu7zauquohshee7yei6koo2phae4quohtioXe3Mahw7Pee4wie4ii9Oht3ooj9naes2kahNg8nei1tahn0Ahm5nipea8Ohw7eec0feenahngeeH2Rupohbae3ur7aitietaePooca1aeghooDohZ8Oof7die2Uosoenae7zie1aighaghaed0pheet0ohg7dohw2EiHe0yap5dou9Ahcohs9eeph4sooriehi","providerId":"ldap","providerType":"org.keycloak.storage.UserStorageProvider","parentId":null,"subType":null,"config":{"pagination":["true"],"fullSyncPeriod":["-1"],"connectionPooling":["true"],"usersDn":["someDN"],"cachePolicy":["DEFAULT"],"useKerberosForPasswordAuthentication":["false"],"importEnabled":["true"],"enabled":["true"],"changedSyncPeriod":["-1"],"bindCredential":["adadasdasdasdasdasdasdasd"],"bindDn":["someBindDN"],"usernameLDAPAttribute":["userPrincipalName"],"lastSync":["1539695370"],"vendor":["ad"],"uuidLDAPAttribute":["objectGUID"],"allowKerberosAuthentication":["false"],"connectionUrl":["ldaps://1.something.local ldaps://1.something.local"],"syncRegistrations":["false"],"authType":["simple"],"connectionTimeout":["60000"],"debug":["false"],"searchScope":["2"],"useTruststoreSpi":["never"],"priority":["0"],"userObjectClasses":["person, organizationalPerson, user"],"rdnLDAPAttribute":["cn"],"readTimeout":["60000"],"editMode":["WRITABLE"],"validatePasswordPolicy":["false"],"batchSizeForSync":["1000"]}}' -v
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> POST /auth/admin/realms/master/components HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.77.0
> Accept: */*
> Authorization: Bearer 
> Content-Type: application/json
> Content-Length: 1307
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 500 Internal Server Error
< Connection: keep-alive
< Content-Type: text/html;charset=UTF-8
< Content-Length: 80
< Date: Sat, 12 Feb 2022 17:54:08 GMT
<
* Connection #0 to host localhost left intact
<html><head><title>Error</title></head><body>Internal Server Error</body></html>%
```

### Actual behavior

Keycloak returns 201 Created, but the component will not created.

curl against `quay.io/keycloak/keycloak:17.0.0`

```
curl http://localhost:8080/admin/realms/master/components -H "Authorization: Bearer" -H "Content-Type: application/json" -d '{"id":null,"name":"em3eedoh8biuph8aiwo0oe2Uyaexu7zauquohshee7yei6koo2phae4quohtioXe3Mahw7Pee4wie4ii9Oht3ooj9naes2kahNg8nei1tahn0Ahm5nipea8Ohw7eec0feenahngeeH2Rupohbae3ur7aitietaePooca1aeghooDohZ8Oof7die2Uosoenae7zie1aighaghaed0pheet0ohg7dohw2EiHe0yap5dou9Ahcohs9eeph4sooriehi","providerId":"ldap","providerType":"org.keycloak.storage.UserStorageProvider","parentId":null,"subType":null,"config":{"pagination":["true"],"fullSyncPeriod":["-1"],"connectionPooling":["true"],"usersDn":["someDN"],"cachePolicy":["DEFAULT"],"useKerberosForPasswordAuthentication":["false"],"importEnabled":["true"],"enabled":["true"],"changedSyncPeriod":["-1"],"bindCredential":["adadasdasdasdasdasdasdasd"],"bindDn":["someBindDN"],"usernameLDAPAttribute":["userPrincipalName"],"lastSync":["1539695370"],"vendor":["ad"],"uuidLDAPAttribute":["objectGUID"],"allowKerberosAuthentication":["false"],"connectionUrl":["ldaps://1.something.local ldaps://1.something.local"],"syncRegistrations":["false"],"authType":["simple"],"connectionTimeout":["60000"],"debug":["false"],"searchScope":["2"],"useTruststoreSpi":["never"],"priority":["0"],"userObjectClasses":["person, organizationalPerson, user"],"rdnLDAPAttribute":["cn"],"readTimeout":["60000"],"editMode":["WRITABLE"],"validatePasswordPolicy":["false"],"batchSizeForSync":["1000"]}}' -v
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> POST /admin/realms/master/components HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.77.0
> Accept: */*
> Authorization: Bearer 
> Content-Type: application/json
> Content-Length: 1307
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 201 Created
< Referrer-Policy: no-referrer
< X-Frame-Options: SAMEORIGIN
< Strict-Transport-Security: max-age=31536000; includeSubDomains
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Location: http://localhost:8080/admin/realms/master/components/24edb19b-c277-45e2-bf52-39a14869917a
< content-length: 0
<
* Connection #0 to host localhost left intact
*
```

### How to Reproduce?

Using cURL

```bash
curl http://localhost:8080/admin/realms/master/components -H "Authorization: Bearer $TOKEN" -H "Content-Type: application/json" -d '{"id":null,"name":"em3eedoh8biuph8aiwo0oe2Uyaexu7zauquohshee7yei6koo2phae4quohtioXe3Mahw7Pee4wie4ii9Oht3ooj9naes2kahNg8nei1tahn0Ahm5nipea8Ohw7eec0feenahngeeH2Rupohbae3ur7aitietaePooca1aeghooDohZ8Oof7die2Uosoenae7zie1aighaghaed0pheet0ohg7dohw2EiHe0yap5dou9Ahcohs9eeph4sooriehi","providerId":"ldap","providerType":"org.keycloak.storage.UserStorageProvider","parentId":null,"subType":null,"config":{"pagination":["true"],"fullSyncPeriod":["-1"],"connectionPooling":["true"],"usersDn":["someDN"],"cachePolicy":["DEFAULT"],"useKerberosForPasswordAuthentication":["false"],"importEnabled":["true"],"enabled":["true"],"changedSyncPeriod":["-1"],"bindCredential":["adadasdasdasdasdasdasdasd"],"bindDn":["someBindDN"],"usernameLDAPAttribute":["userPrincipalName"],"lastSync":["1539695370"],"vendor":["ad"],"uuidLDAPAttribute":["objectGUID"],"allowKerberosAuthentication":["false"],"connectionUrl":["ldaps://1.something.local ldaps://1.something.local"],"syncRegistrations":["false"],"authType":["simple"],"connectionTimeout":["60000"],"debug":["false"],"searchScope":["2"],"useTruststoreSpi":["never"],"priority":["0"],"userObjectClasses":["person, organizationalPerson, user"],"rdnLDAPAttribute":["cn"],"readTimeout":["60000"],"editMode":["WRITABLE"],"validatePasswordPolicy":["false"],"batchSizeForSync":["1000"]}}'
```

### Anything else?

Logs from Keycloak

```
2022-02-12 17:44:01,062 WARN  [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (executor-thread-0) SQL Error: 22001, SQLState: 22001
2022-02-12 17:44:01,064 ERROR [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (executor-thread-0) Value too long for column "NAME VARCHAR(255)": "'em3eedoh8biuph8aiwo0oe2Uyaexu7zauquohshee7yei6koo2phae4quohtioXe3Mahw7Pee4wie4ii9Oht3ooj9naes2kahNg8nei1tahn0Ahm5nipea8Ohw7eec0... (256)"; SQL statement:
insert into COMPONENT (NAME, PARENT_ID, PROVIDER_ID, PROVIDER_TYPE, REALM_ID, SUB_TYPE, ID) values (?, ?, ?, ?, ?, ?, ?) [22001-197]
2022-02-12 17:44:01,069 INFO  [org.hibernate.engine.jdbc.batch.internal.AbstractBatchImpl] (executor-thread-0) HHH000010: On release of batch it still contained JDBC statements
2022-02-12 17:44:01,080 WARN  [com.arjuna.ats.arjuna] (executor-thread-0) ARJUNA012125: TwoPhaseCoordinator.beforeCompletion - failed for SynchronizationImple< 0:ffffac110003:a63b:6207f1d7:52, org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization@4167c22b >: javax.persistence.PersistenceException: org.hibernate.exception.DataException: could not execute statement
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:154)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:188)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1406)
	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:493)
	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3285)
	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2420)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:449)
	at org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl.beforeCompletion(JtaTransactionCoordinatorImpl.java:356)
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl.beforeCompletion(SynchronizationCallbackCoordinatorNonTrackingImpl.java:47)
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization.beforeCompletion(RegisteredSynchronization.java:37)
	at com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple.beforeCompletion(SynchronizationImple.java:76)
	at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.beforeCompletion(TwoPhaseCoordinator.java:360)
	at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.end(TwoPhaseCoordinator.java:91)
	at com.arjuna.ats.arjuna.AtomicAction.commit(AtomicAction.java:162)
	at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1295)
	at com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction.commit(BaseTransaction.java:128)
	at io.quarkus.narayana.jta.runtime.CDIDelegatingTransactionManager.commit(CDIDelegatingTransactionManager.java:105)
	at org.keycloak.transaction.JtaTransactionWrapper.commit(JtaTransactionWrapper.java:90)
	at org.keycloak.services.DefaultKeycloakTransactionManager.commit(DefaultKeycloakTransactionManager.java:136)
	at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.close(QuarkusRequestFilter.java:112)
	at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$configureEndHandler$2(QuarkusRequestFilter.java:90)
	at io.vertx.ext.web.impl.RoutingContextImpl.lambda$null$0(RoutingContextImpl.java:545)
	at io.vertx.ext.web.impl.SparseArray.forEachInReverseOrder(SparseArray.java:40)
	at io.vertx.ext.web.impl.RoutingContextImpl.lambda$getHeadersEndHandlers$1(RoutingContextImpl.java:545)
	at io.vertx.core.http.impl.Http1xServerResponse.prepareHeaders(Http1xServerResponse.java:707)
	at io.vertx.core.http.impl.Http1xServerResponse.end(Http1xServerResponse.java:411)
	at io.vertx.core.http.impl.Http1xServerResponse.end(Http1xServerResponse.java:391)
	at io.vertx.core.http.impl.Http1xServerResponse.end(Http1xServerResponse.java:460)
	at io.quarkus.resteasy.runtime.standalone.VertxBlockingOutput.write(VertxBlockingOutput.java:71)
	at io.quarkus.resteasy.runtime.standalone.VertxHttpResponse.writeBlocking(VertxHttpResponse.java:172)
	at io.quarkus.resteasy.runtime.standalone.VertxOutputStream.close(VertxOutputStream.java:126)
	at io.quarkus.resteasy.runtime.standalone.VertxHttpResponse.finish(VertxHttpResponse.java:145)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:170)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:82)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:42)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:67)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:55)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$5.handle(VertxHttpRecorder.java:362)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$5.handle(VertxHttpRecorder.java:340)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
	at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$createBlockingHandler$1(QuarkusRequestFilter.java:66)
	at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:159)
	at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:157)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.hibernate.exception.DataException: could not execute statement
	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:115)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:37)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:200)
	at org.hibernate.engine.jdbc.batch.internal.NonBatchingBatch.addToBatch(NonBatchingBatch.java:46)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3375)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3908)
	at org.hibernate.action.internal.EntityInsertAction.execute(EntityInsertAction.java:107)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604)
	at org.hibernate.engine.spi.ActionQueue.lambda$executeActions$1(ActionQueue.java:478)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:475)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:344)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:40)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:107)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1402)
	... 56 more
Caused by: org.h2.jdbc.JdbcSQLException: Value too long for column "NAME VARCHAR(255)": "'em3eedoh8biuph8aiwo0oe2Uyaexu7zauquohshee7yei6koo2phae4quohtioXe3Mahw7Pee4wie4ii9Oht3ooj9naes2kahNg8nei1tahn0Ahm5nipea8Ohw7eec0... (256)"; SQL statement:
insert into COMPONENT (NAME, PARENT_ID, PROVIDER_ID, PROVIDER_TYPE, REALM_ID, SUB_TYPE, ID) values (?, ?, ?, ?, ?, ?, ?) [22001-197]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:357)
	at org.h2.message.DbException.get(DbException.java:179)
	at org.h2.table.Column.validateConvertUpdateSequence(Column.java:398)
	at org.h2.table.Table.validateConvertUpdateSequence(Table.java:798)
	at org.h2.command.dml.Insert.insertRows(Insert.java:177)
	at org.h2.command.dml.Insert.update(Insert.java:134)
	at org.h2.command.CommandContainer.update(CommandContainer.java:102)
	at org.h2.command.Command.executeUpdate(Command.java:261)
	at org.h2.jdbc.JdbcPreparedStatement.executeUpdateInternal(JdbcPreparedStatement.java:199)
	at org.h2.jdbc.JdbcPreparedStatement.executeUpdate(JdbcPreparedStatement.java:153)
	at io.agroal.pool.wrapper.PreparedStatementWrapper.executeUpdate(PreparedStatementWrapper.java:88)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:197)
	... 68 more
```

--- a/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/web/QuarkusRequestFilter.java
+++ b/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/web/QuarkusRequestFilter.java
@@ -25,6 +25,9 @@ import org.keycloak.models.KeycloakSession;
 import org.keycloak.models.KeycloakSessionFactory;
 import org.keycloak.models.KeycloakTransactionManager;
 
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.handler.codec.http.HttpResponseStatus;
 import io.vertx.core.AsyncResult;
 import io.vertx.core.Handler;
 import io.vertx.core.Promise;
@@ -64,7 +67,7 @@ public class QuarkusRequestFilter implements Handler<RoutingContext> {
             try {
                 tx.begin();
                 context.next();
-                promise.complete();
+                promise.tryComplete();
             } catch (Throwable cause) {
                 promise.fail(cause);
                 // re-throw so that the any exception is handled from parent
@@ -90,6 +93,10 @@ public class QuarkusRequestFilter implements Handler<RoutingContext> {
                 close(session);
             } catch (Throwable cause) {
                 promise.fail(cause);
+                context.response().headers().clear();
+                context.response().putHeader(HttpHeaderNames.CONTENT_TYPE, HttpHeaderValues.TEXT_PLAIN);
+                context.response().putHeader(HttpHeaderNames.CONTENT_LENGTH, "0");
+                context.response().setStatusCode(HttpResponseStatus.INTERNAL_SERVER_ERROR.code());
             }
         });
     }
@@ -105,14 +112,18 @@ public class QuarkusRequestFilter implements Handler<RoutingContext> {
 
     protected void close(KeycloakSession session) {
         KeycloakTransactionManager tx = session.getTransactionManager();
-        if (tx.isActive()) {
-            if (tx.getRollbackOnly()) {
-                tx.rollback();
-            } else {
-                tx.commit();
+
+        try {
+            if (tx.isActive()) {
+                if (tx.getRollbackOnly()) {
+                    tx.rollback();
+                } else {
+                    tx.commit();
+                }
             }
+        } finally {
+            session.close();
         }
-        session.close();
     }
 
     private ClientConnection createClientConnection(HttpServerRequest request) {

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/ComponentsTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/ComponentsTest.java
@@ -173,6 +173,34 @@ public class ComponentsTest extends AbstractAdminTest {
     }
 
     @Test
+    public void failCreateWithLongName() {
+        StringBuilder name = new StringBuilder();
+
+        while (name.length() < 30) {
+            name.append("invalid");
+        }
+
+        ComponentRepresentation rep = createComponentRepresentation(name.toString());
+
+        rep.getConfig().addFirst("required", "foo");
+
+        ComponentsResource components = realm.components();
+
+        try (Response response = components.add(rep)) {
+            if (Response.Status.INTERNAL_SERVER_ERROR.getStatusCode() == response.getStatus()) {
+                // using database should fail due to constraint violations
+                assertFalse(components.query().stream().map(ComponentRepresentation::getName).anyMatch(name.toString()::equals));
+            } else if (Response.Status.CREATED.getStatusCode() == response.getStatus()) {
+                // using the map storage should work because there are no constraints
+                String id = ApiUtil.getCreatedId(response);
+                assertNotNull(components.component(id).toRepresentation());
+            } else {
+                fail("Unexpected response");
+            }
+        }
+    }
+
+    @Test
     public void testUpdate() {
         ComponentRepresentation rep = createComponentRepresentation("mycomponent");
 

